### PR TITLE
X prov map

### PR DESCRIPTION
**Pull request by natjoe4**
# Summary
Added functionality for mapping X-Provenance headers to an AuditEvent resource and storing that resource in our mongo db for $submit-data requests

## New behavior
When sending a submit-data request, if you add the optional X-Provenance header, an AuditEvent will be created with the appropriate properties, documenting the action that occurred and the agent that was responsible

## Code changes
- Added new provenanceUtils.js file with function for mapping headers to AuditEvent Object
- Editted the existing submitData function to create the AuditEvent resource when necessary
- Added testing and fixtures for new functions

# Testing guidance
- Run all tests
- Connect to robomongo
- Try sending a `$submit-data` request (more on that [here](https://github.com/projecttacoma/deqm-test-server/pull/7)) with X-Provenance headers. I'll post a couple of options in Slack which have worked for me
- Ensure that AuditEvent resources are created correctly displaying the information stored in the X-Provenance Header
- Here's the [AuditEvent IG](http://www.hl7.org/fhir/auditevent.html). Please look through the documentation and ensure that all fields are of the correct cardinality and are mapped correctly.

### CODE REVIEW

File: src/util/provenanceUtils.js
Comment (sarahmcdougall) [reply src/util/provenanceUtils.js:10]: 0 represents success and C represents "Create" but I didn't know that without looking at the IG

File: src/util/provenanceUtils.js
Comment (sarahmcdougall) [inline src/util/provenanceUtils.js:34]: According to the IG, there must be at least one audit agent and the agent must have a requestor. Will there ever be a case where the provenance does not have the `agent` attribute? In that case, I don't think the `audit.agent` will get a requestor because it won't go through the if statement `if (provenance.agent)`

File: src/util/provenanceUtils.js
Comment (sarahmcdougall) [inline src/util/provenanceUtils.js:6]:
```suggestion
 * @param {string} provenance the provenance headers in string form from the request
```

File: src/util/provenanceUtils.js
Comment (sarahmcdougall) [inline src/util/provenanceUtils.js:7]:
```suggestion
 * @param {Object} args the object containing the base_version parameter
```

File: src/util/provenanceUtils.js
Comment (sarahmcdougall) [inline src/util/provenanceUtils.js:63]:
```suggestion
 * @param {Object} reference a fhir reference object to the delegating agent
```

**Review by sarahmcdougall — CHANGES_REQUESTED** [review]
Functionality seems to be what we want - I was able to send a `$submit-data` request with a provenance header and the AuditEvent resource was created with the necessary information. Unit tests pass as well. I have some small changes and suggestions on where to add comments/more clarification. There are a few spots where I'm not entirely sure if the code matches the AuditEvent IG. As we mentioned earlier, I think it would be helpful to get a second reviewer on this so that we have another set of eyes on the AuditEvent IG.

File: test/fixtures/testProvenanceUtils.js
Comment (mgramigna) [inline test/fixtures/testProvenanceUtils.js:45]:
```suggestion
          code: 'HQUALIMP',
          display: 'health quality improvement'
```

The code we want here is the [health quality improvement one](https://www.hl7.org/fhir/v3/ActReason/cs.html#v3-ActReason-HQUALIMP), not care management. It looks like there's no hardcoding of this so that's good, you'll just have to update your test fixtures. I'll make suggestions to make your life easier and you can batch them

File: test/fixtures/testProvenanceUtils.js
Comment (mgramigna) [inline test/fixtures/testProvenanceUtils.js:101]:
```suggestion
          code: 'HQUALIMP',
          display: 'health quality improvement'
```

File: test/fixtures/testProvenanceUtils.js
Comment (mgramigna) [inline test/fixtures/testProvenanceUtils.js:181]:
```suggestion
          code: 'HQUALIMP',
          display: 'health quality improvement'
```

File: test/fixtures/testProvenanceUtils.js
Comment (mgramigna) [inline test/fixtures/testProvenanceUtils.js:248]:
```suggestion
          code: 'HQUALIMP',
          display: 'health quality improvement'
```

File: test/fixtures/testProvenanceUtils.js
Comment (mgramigna) [inline test/fixtures/testProvenanceUtils.js:309]:
```suggestion
          code: 'HQUALIMP',
          display: 'health quality improvement'
```

File: test/fixtures/testProvenanceUtils.js
Comment (mgramigna) [inline test/fixtures/testProvenanceUtils.js:395]:
```suggestion
          code: 'HQUALIMP',
          display: 'health quality improvement'
```

File: src/util/provenanceUtils.js
Comment (mgramigna) [reply src/util/provenanceUtils.js:38]: I think this will kind of depend. Since `agent` is `1..*` in provenance as well, theoretically we don't have to worry about Provenance missing it for the case of AuditEvent creation. However, it's TBD still on what we want to do if we receive an X-Provenance that isn't valid. Do we still create an AuditEvent that says that a submission failed because of an invalid Provenance? Maybe? So I think future tasking could be to add provenance validation to at least ensure that the required properties are there. If they're not, maybe we create an AuditEvent for a failed submission and can update the outcome accordingly.

For this task, I think it's fine to just assume that the provenance you're getting is good for a proper mapping, and we can worry about enforcing stuff later.

File: src/util/provenanceUtils.js
Comment (mgramigna) [inline src/util/provenanceUtils.js:43]:
```suggestion
      if (agent.role.coding.some(role => role.code === 'AGNT')) {
```

nitpick: this function does the same thing but intuitively returns a boolean rather than checking the length of a filtered array

File: src/util/provenanceUtils.js
Comment (mgramigna) [inline src/util/provenanceUtils.js:47]:
```suggestion
      if (agent.role.coding.some(role => role.code === 'ASSIGNED')) {
```

picking nits

File: src/util/provenanceUtils.js
Comment (mgramigna) [inline src/util/provenanceUtils.js:55]: Might be better to just use `delete agent.onBehalfOf` here so the property disappears completely rather than being set to a nullish value. I would give this a try and see if it gives you any undesired side effects

File: src/util/provenanceUtils.js
Comment (mgramigna) [reply src/util/provenanceUtils.js:38]: EDIT: I see we have a helper function for validating provenance down below. If you want to just quickly enforce the 1..*-ness down there, that's cool. I think the future tasking that I'm talking about would only consider if we still want to create AuditEvents for bad provenance input, which can be scoped to a diff task

File: src/util/provenanceUtils.js
Comment (mgramigna) [inline src/util/provenanceUtils.js:91]: Placeholder: @ssayer had some thoughts about the use of `DELEGATOR` here that we are still working through. Stay tuned

File: test/provenanceUtils.test.js
Comment (mgramigna) [inline test/provenanceUtils.test.js:1]:
```suggestion
const { resolveSchema } = require('@projecttacoma/node-fhir-server-core');
```

I believe this is exported by the root of `node-fhir-server-core`

File: test/provenanceUtils.test.js
Comment (mgramigna) [inline test/provenanceUtils.test.js:19]: clutch

File: test/provenanceUtils.test.js
Comment (mgramigna) [inline test/provenanceUtils.test.js:25]: clutch

**Review by mgramigna — COMMENTED** [review]
Made a bunch of inline comments. Some need more conversation that hopefully @ssayer can talk to tomorrow morning, but this should at least give you some actions to do on this branch if you're working tomorrow.

File: src/util/provenanceUtils.js
Comment (natjoe4) [reply src/util/provenanceUtils.js:43]: So useful. Didn't know about this one. Thanks!

File: src/util/provenanceUtils.js
Comment (natjoe4) [reply src/util/provenanceUtils.js:22]: I think we ultimately don't. I added a TODO here, but not sure how we want to map the success of the submission to a code here

File: src/services/base.service.js
Comment (sarahmcdougall) [inline src/services/base.service.js:285]: Thank you for moving this :)

File: src/util/provenanceUtils.js
Comment (sarahmcdougall) [inline src/util/provenanceUtils.js:52]:
```suggestion
      // Delete onBehalfOf since we already store this info as
```

File: src/util/provenanceUtils.js
Comment (sarahmcdougall) [inline src/util/provenanceUtils.js:93]:
```suggestion
 * @param {Object} requestHeaders the headers from the request body
```

**Review by sarahmcdougall — CHANGES_REQUESTED** [review]
I'm not an AuditEvent/Provenance expert, but I think this is pretty good. Functionality is still good and the updated unit tests pass. You made all the changes that we asked for last week and so I think this work aligns better with the IG. I have the tiniest of comments for documentation but that's it. Let me know if there's anything you want a closer look at and I can do a deeper dive into the IG.

**Review by mgramigna — APPROVED** [review]
¯\_(ツ)_/¯

**Review by sarahmcdougall — APPROVED** [review]
¯_(ツ)_/¯

**Pull request merged**
